python/ray/air/util/tensor_extensions/arrow.py

```diff
@@ -124,10 +127,53 @@ def _extension_scalar_to_ndarray(
             """
             Convert an ExtensionScalar to a tensor element.
             """
-            # TODO(Clark): Construct ndarray view directly on tensor element buffer to
-            # ensure reliable zero-copy semantics.
-            flat_ndarray = scalar.value.values.to_numpy(zero_copy_only=False)
-            return flat_ndarray.reshape(self.shape)
+            shape = scalar.type.shape
+            raw_values = scalar.value.values
+            offset = raw_values.offset
+            if pa.types.is_boolean(raw_values.type):
+                # Arrow boolean array buffers are bit-packed, with 8 entries per byte,
+                # and are accessed via bit offsets.
+                buffer_item_width = raw_values.type.bit_width
+            else:
+                # We assume all other array types are accessed via byte array
+                # offsets.
+                buffer_item_width = raw_values.type.bit_width // 8
+            data_buffer = raw_values.buffers()[1]
+            data_offset = buffer_item_width * offset
+            if pa.types.is_boolean(raw_values.type):
+                # Special handling for boolean arrays, since Arrow bit-packs
+                # boolean arrays while NumPy does not.
+                # Cast as uint8 array and let NumPy unpack into a boolean view.
+                # Offset into uint8 array, where each element is
+                # a bucket for 8 booleans.
+                byte_bucket_offset = offset // 8
+                # Offset for a specific boolean, within a uint8 array element.
+                bool_offset = offset % 8
+                # The number of uint8 array elements (buckets) that our slice spans.
+                # Note that, due to the offset for a specific boolean,
+                # the slice can span byte boundaries even if it contains
+                # less than 8 booleans.
+                num_boolean_byte_buckets = 1 + ((bool_offset + np.prod(shape) - 1) // 8)
+                # Construct the uint8 array view on the buffer.
+                arr = np.ndarray(
+                    (num_boolean_byte_buckets,),
+                    dtype=np.uint8,
+                    buffer=data_buffer,
+                    offset=byte_bucket_offset,
+                )
+                # Unpack into a byte per boolean, using LSB bit-packed ordering.
+                arr = np.unpackbits(arr, bitorder="little")
+                # Interpret buffer as boolean array.
+                return np.ndarray(shape, dtype=np.bool_, buffer=arr, offset=bool_offset)
+            ext_dtype = raw_values.type.to_pandas_dtype()
+            # Special handling of ragged string tensors
+            if pa.types.is_fixed_size_binary(raw_values.type):
+                ext_dtype = np.dtype(
+                    f"<U{raw_values.type.byte_width // NUM_BYTES_PER_UNICODE_CHAR}"
+                )
+            return np.ndarray(
+                shape, dtype=ext_dtype, buffer=data_buffer, offset=data_offset
+            )
```

Comment:
Btw, it appears that this and `ArrowVariableShapedTensorType._extension_scalar_to_ndarray()` are the same after the definitions of `shape` and `raw_values`, could we consolidate most of this logic into a single utility function that lives at the bottom of this module?

Reply:
I generalized `ArrowTensorType._extension_scalar_to_ndarray` as well (which had a few variables that were slightly different). LMK what you think